### PR TITLE
Translation layer auth

### PR DESCRIPTION
**Pull request by angrykoala**
# Description

Deals with most auth

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3798.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

File: packages/graphql/src/schema-model/relationship/Relationship.ts [hunk 1]
```diff
@@ -35,6 +35,7 @@ export class Relationship {
     public readonly target: Entity;
     public readonly direction: RelationshipDirection;
     public readonly isArray: boolean;
+    public readonly isNullable: boolean;
```
Comment (a-alle) [inline packages/graphql/src/schema-model/relationship/Relationship.ts:38]: together with isArray it's replicating the AttributeType at this point (which also supports Object, Union, Interface) - also what if it's an array of non-nullable items? 
Maybe the current `type` property should be renamed to `relationshipType` and actual `type` should be an AttributeType. Thoughts?

Comment (MacondoExpress) [reply packages/graphql/src/schema-model/relationship/Relationship.ts:38]: I agree the AttributeType structure can become handy here! But I think should not be relevant for the scope of this PR and rather be part of some Schema Model improvements PR!

File: packages/graphql/src/translate/queryAST/ast/QueryASTContext.ts [hunk 2]
```diff
@@ -37,4 +37,12 @@ export class QueryASTContext {
         this.relationship = relationship;
         this.source = source;
     }
+
+    public push({ relationship, target }: { relationship: Cypher.Relationship; target: Cypher.Node }): QueryASTContext {
```
Comment (MacondoExpress) [inline packages/graphql/src/translate/queryAST/ast/QueryASTContext.ts:41]: I think here push here it's a little bit ambiguous as it seems that we push a new Context in some stack when in reality we just return a new QueryASTContext

File: packages/graphql/src/translate/queryAST/ast/filters/RelationshipFilter.ts [hunk 3]
```diff
@@ -25,10 +25,10 @@ import { QueryASTContext } from "../QueryASTContext";
 import type { RelationshipAdapter } from "../../../../schema-model/relationship/model-adapters/RelationshipAdapter";
 
 export class RelationshipFilter extends Filter {
-    private targetNodeFilters: Filter[] = [];
-    private relationship: RelationshipAdapter;
-    private operator: RelationshipWhereOperator;
-    private isNot: boolean;
+    protected targetNodeFilters: Filter[] = [];
```
Comment (MacondoExpress) [inline packages/graphql/src/translate/queryAST/ast/filters/RelationshipFilter.ts:28]: On the ConnectionFilter this property is named `innerFilters` not that it's a better name but we could choose a consistent name for both!

File: packages/graphql/src/translate/queryAST/ast/filters/authorization-filters/AuthRelationshipFilter.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            return [];
+        }
+        return [new Cypher.OptionalMatch(pattern).with("*", [Cypher.count(nestedContext.target), this.countVar])];
+    }
+
+    public getPredicate(queryASTContext: QueryASTContext): Cypher.Predicate | undefined {
+        // if version>5 super.getPredicate()
+
+        //
+        const nestedContext = queryASTContext.push({
```
Comment (MacondoExpress) [inline packages/graphql/src/translate/queryAST/ast/filters/authorization-filters/AuthRelationshipFilter.ts:51]: Nice, we should start adopting as soon as possible this mechanism everywhere!

File: packages/graphql/tests/tck/directives/authorization/arguments/allow/allow.test.ts [hunk 5]
```diff
@@ -198,13 +198,13 @@ describe("Cypher Auth Allow", () => {
                 WITH this
                 MATCH (this)-[this0:HAS_POST]->(this1:Post)
                 OPTIONAL MATCH (this1)<-[:HAS_POST]-(this2:User)
-                WITH *, count(this2) AS creatorCount
+                WITH *, count(this2) AS var3
```
Comment (MacondoExpress) [inline packages/graphql/tests/tck/directives/authorization/arguments/allow/allow.test.ts:201]: always lovely to see a Cypher.Variable introduced!

**Review by MacondoExpress — APPROVED** [review]
Neat!

File: packages/graphql/src/schema-model/relationship/Relationship.ts (hunk 1 above)
Comment (a-alle) [reply packages/graphql/src/schema-model/relationship/Relationship.ts:38]: yes, I meant it as part of the Schema Model changes, I agree not in the scope of this PR

File: packages/graphql/src/translate/queryAST/ast/filters/RelationshipFilter.ts (hunk 3 above)
Comment (angrykoala) [reply packages/graphql/src/translate/queryAST/ast/filters/RelationshipFilter.ts:28]: Agree, I'll review all filters holistically to find common naming

File: packages/graphql/src/translate/queryAST/ast/QueryASTContext.ts (hunk 2 above)
Comment (angrykoala) [reply packages/graphql/src/translate/queryAST/ast/QueryASTContext.ts:41]: :thinking: 

The idea was kind of ("push a new relationship into the stack"), but maybe it is not needed and we can find a better way

**Pull request merged**